File: backend/services/naming.py
"""Auto-naming engine for campaigns, ad sets, and ads."""
from datetime import datetime
from typing import Optional
import re
import string

# Default naming templates
DEFAULT_TEMPLATES = {
    "campaign": "{objective}_{budget_type}_{name}_{date}",
    "adset": "{campaign_short}_{adset_name}_{optimization}_{variant}",
    "ad": "{adset_short}_{creative_ref}_{format}_{version}",
}


def _shorten(name: str, max_len: int = 15) -> str:
    """Create an abbreviated version of a name."""
    # Remove common words
    stop_words = {"de", "del", "la", "el", "los", "las", "the", "a", "an", "for", "and"}
    words = name.split()
    words = [w for w in words if w.lower() not in stop_words]

    if not words:
        return name[:max_len]

    # Try initials + first word
    if len(words) == 1:
        return words[0][:max_len]

    # CamelCase abbreviation
    abbr = "".join(w[0].upper() + w[1:3] for w in words[:4])
    return abbr[:max_len]


def _clean_name(name: str) -> str:
    """Clean a name for use in Meta (remove special chars)."""
    # Replace spaces with underscores, keep alphanumeric and some special chars
    cleaned = re.sub(r"[^a-zA-Z0-9_\-.]", "_", name)
    cleaned = re.sub(r"_+", "_", cleaned)  # collapse multiple underscores
    return cleaned.strip("_")


def generate_campaign_name(
    name: str,
    objective: str,
    budget_type: str,
    template: Optional[str] = None,
) -> str:
    """Generate campaign name from template."""
    tmpl = template or DEFAULT_TEMPLATES["campaign"]
    obj_short = objective.replace("OUTCOME_", "")
    date = datetime.now().strftime("%Y%m%d")

    result = tmpl.format(
        name=name,
        objective=obj_short,
        budget_type=budget_type,
        date=date,
    )
    return _clean_name(result)


def generate_adset_name(
    campaign_name: str,
    adset_name: str,
    optimization_goal: str,
    variant_index: int,
    template: Optional[str] = None,
) -> str:
    """Generate ad set name from template."""
    tmpl = template or DEFAULT_TEMPLATES["adset"]
    campaign_short = _shorten(campaign_name)

    # Variant letter: A, B, C, ...
    variant = string.ascii_uppercase[variant_index % 26]

    result = tmpl.format(
        campaign_short=campaign_short,
        adset_name=_shorten(adset_name),
        optimization=optimization_goal.replace("OFFSITE_", "").replace("_", "")[:8],
        variant=variant,
    )
    return _clean_name(result)


def generate_ad_name(
    adset_name: str,
    creative_ref: str,
    creative_format: str,
    version_index: int,
    template: Optional[str] = None,
) -> str:
    """Generate ad name from template."""
    tmpl = template or DEFAULT_TEMPLATES["ad"]
    adset_short = _shorten(adset_name)

    result = tmpl.format(
        adset_short=adset_short,
        creative_ref=creative_ref[:20],
        format=creative_format[:5] if creative_format else "multi",
        version=f"v{version_index + 1}",
    )
    return _clean_name(result)


def generate_all_names(
    project_name: str,
    objective: str,
    budget_type: str,
    ad_sets: list[dict],
    templates: Optional[dict] = None,
) -> dict:
    """Generate all names for a complete campaign structure.

    Returns: {
        "campaign_name": "...",
        "ad_sets": [
            {"generated_name": "...", "ads": [{"generated_name": "..."}]}
        ]
    }
    """
    campaign_tmpl = (templates or {}).get("campaign")
    adset_tmpl = (templates or {}).get("adset")
    ad_tmpl = (templates or {}).get("ad")

    campaign_name = generate_campaign_name(project_name, objective, budget_type, campaign_tmpl)

    result_sets = []
    for i, adset in enumerate(ad_sets):
        adset_gen_name = generate_adset_name(
            campaign_name, adset["name"], adset.get("optimization_goal", "CONVERSIONS"), i, adset_tmpl
        )
        ads = []
        for j, ad in enumerate(adset.get("ads", [])):
            ad_gen_name = generate_ad_name(
                adset_gen_name, ad.get("creative_ref", "creative"), ad.get("format", ""), j, ad_tmpl
            )
            ads.append({"generated_name": ad_gen_name, **ad})
        result_sets.append({"generated_name": adset_gen_name, "ads": ads, **adset})

    return {
        "campaign_name": campaign_name,
        "ad_sets": result_sets,
    }

File: api/lib/services/naming.ts
/**
 * Auto-naming engine for campaigns, ad sets, and ads.
 */

// ── Default naming templates ────────────────────────────────────────────────

const DEFAULT_TEMPLATES = {
  campaign: "{objective}_{budget_type}_{name}_{date}",
  adset: "{campaign_short}_{adset_name}_{optimization}_{variant}",
  ad: "{adset_short}_{creative_ref}_{format}_{version}",
} as const;

// ── Internal helpers ────────────────────────────────────────────────────────

function shorten(name: string, maxLen: number = 15): string {
  const stopWords = new Set([
    "de", "del", "la", "el", "los", "las", "the", "a", "an", "for", "and",
  ]);
  const words = name.split(/\s+/).filter((w) => !stopWords.has(w.toLowerCase()));

  if (words.length === 0) {
    return name.slice(0, maxLen);
  }

  if (words.length === 1) {
    return words[0].slice(0, maxLen);
  }

  // CamelCase abbreviation
  const abbr = words
    .slice(0, 4)
    .map((w) => w[0].toUpperCase() + w.slice(1, 3))
    .join("");
  return abbr.slice(0, maxLen);
}

function cleanName(name: string): string {
  // Replace non-alphanumeric (except _ - .) with underscores
  let cleaned = name.replace(/[^a-zA-Z0-9_\-.]/g, "_");
  // Collapse multiple underscores
  cleaned = cleaned.replace(/_+/g, "_");
  // Strip leading/trailing underscores
  cleaned = cleaned.replace(/^_+|_+$/g, "");
  return cleaned;
}

function formatTemplate(
  template: string,
  values: Record<string, string>,
): string {
  let result = template;
  for (const [key, value] of Object.entries(values)) {
    result = result.replace(new RegExp(`\\{${key}\\}`, "g"), value);
  }
  return result;
}

// ── Public API ──────────────────────────────────────────────────────────────

export function generateCampaignName(
  name: string,
  objective: string,
  budgetType: string,
  template?: string,
): string {
  const tmpl = template ?? DEFAULT_TEMPLATES.campaign;
  const objShort = objective.replace("OUTCOME_", "");
  const now = new Date();
  const date =
    String(now.getFullYear()) +
    String(now.getMonth() + 1).padStart(2, "0") +
    String(now.getDate()).padStart(2, "0");

  const result = formatTemplate(tmpl, {
    name,
    objective: objShort,
    budget_type: budgetType,
    date,
  });
  return cleanName(result);
}

export function generateAdSetName(
  campaignName: string,
  adsetName: string,
  optimizationGoal: string,
  variantIndex: number,
  template?: string,
): string {
  const tmpl = template ?? DEFAULT_TEMPLATES.adset;
  const campaignShort = shorten(campaignName);

  // Variant letter: A, B, C, ...
  const variant = String.fromCharCode(65 + (variantIndex % 26));

  const optShort = optimizationGoal
    .replace("OFFSITE_", "")
    .replace(/_/g, "")
    .slice(0, 8);

  const result = formatTemplate(tmpl, {
    campaign_short: campaignShort,
    adset_name: shorten(adsetName),
    optimization: optShort,
    variant,
  });
  return cleanName(result);
}

export function generateAdName(
  adsetName: string,
  creativeRef: string,
  creativeFormat: string,
  versionIndex: number,
  template?: string,
): string {
  const tmpl = template ?? DEFAULT_TEMPLATES.ad;
  const adsetShort = shorten(adsetName);

  const result = formatTemplate(tmpl, {
    adset_short: adsetShort,
    creative_ref: creativeRef.slice(0, 20),
    format: creativeFormat ? creativeFormat.slice(0, 5) : "multi",
    version: `v${versionIndex + 1}`,
  });
  return cleanName(result);
}

export interface AdSetInput {
  name: string;
  optimization_goal?: string;
  ads?: Array<{
    creative_ref?: string;
    format?: string;
    [key: string]: unknown;
  }>;
  [key: string]: unknown;
}

export interface GeneratedNames {
  campaign_name: string;
  ad_sets: Array<{
    generated_name: string;
    ads: Array<{ generated_name: string; [key: string]: unknown }>;
    [key: string]: unknown;
  }>;
}

export function generateAllNames(
  projectName: string,
  objective: string,
  budgetType: string,
  adSets: AdSetInput[],
  templates?: { campaign?: string; adset?: string; ad?: string },
): GeneratedNames {
  const campaignTmpl = templates?.campaign;
  const adsetTmpl = templates?.adset;
  const adTmpl = templates?.ad;

  const campaignName = generateCampaignName(
    projectName,
    objective,
    budgetType,
    campaignTmpl,
  );

  const resultSets = adSets.map((adset, i) => {
    const adsetGenName = generateAdSetName(
      campaignName,
      adset.name,
      adset.optimization_goal ?? "CONVERSIONS",
      i,
      adsetTmpl,
    );

    const generatedAds = (adset.ads ?? []).map((ad, j) => {
      const adGenName = generateAdName(
        adsetGenName,
        ad.creative_ref ?? "creative",
        ad.format ?? "",
        j,
        adTmpl,
      );
      return { ...ad, generated_name: adGenName } as {
        generated_name: string;
        [key: string]: unknown;
      };
    });

    return {
      ...adset,
      generated_name: adsetGenName,
      ads: generatedAds,
    } as {
      generated_name: string;
      ads: Array<{ generated_name: string; [key: string]: unknown }>;
      [key: string]: unknown;
    };
  });

  return {
    campaign_name: campaignName,
    ad_sets: resultSets,
  };
}
